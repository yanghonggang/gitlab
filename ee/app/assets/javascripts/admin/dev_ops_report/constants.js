import { s__ } from '~/locale';

export const DEVOPS_ADOPTION_STRINGS = {
  emptyState: {
    title: s__('DevopsAdoption|Add a segment to get started'),
    description: s__(
      'DevopsAdoption|DevOps adoption uses segments to track adoption across key features. Segments are a way to track multiple related projects and groups at once. For example, you could create a segment for the engineering department or a particular product team.',
    ),
    button: s__('DevopsAdoption|Add new segment'),
  },
};
