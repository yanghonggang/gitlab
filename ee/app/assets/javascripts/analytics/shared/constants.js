import { masks } from 'dateformat';

const { isoDate, mediumDate } = masks;

export const dateFormats = {
  isoDate,
  defaultDate: mediumDate,
  defaultDateTime: 'mmm d, yyyy h:MMtt',
};

export const scatterChartLineProps = {
  default: {
    type: 'line',
    showSymbol: false,
    // By default zlevel is 2 for all series types.
    // By increasing the zlevel to 3 we make sure that the trendline gets drawn in front of the dots in the chart.
    zlevel: 3,
  },
};

export const DATE_RANGE_LIMIT = 180;

export const OFFSET_DATE_BY_ONE = 1;

export const NO_DRAG_CLASS = 'no-drag';

export const DATA_REFETCH_DELAY = 250;
