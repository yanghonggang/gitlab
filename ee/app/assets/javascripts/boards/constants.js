export const DRAGGABLE_TAG = 'div';

/* eslint-disable @gitlab/require-i18n-strings */
export const EpicFilterType = {
  any: 'Any',
  none: 'None',
};

export const IterationFilterType = {
  any: 'Any',
  none: 'None',
};

export const GroupByParamType = {
  epic: 'epic',
};

export default {
  DRAGGABLE_TAG,
  EpicFilterType,
};
