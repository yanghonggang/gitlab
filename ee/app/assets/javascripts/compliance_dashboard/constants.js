export const PRESENTABLE_APPROVERS_LIMIT = 2;

export const COMPLIANCE_TAB_COOKIE_KEY = 'compliance_dashboard_tabs';

export const INPUT_DEBOUNCE = 500;

export const CUSTODY_REPORT_PARAMETER = 'commit_sha';
