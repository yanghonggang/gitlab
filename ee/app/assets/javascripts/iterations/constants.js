export const Namespace = {
  Group: 'group',
  Project: 'project',
};
