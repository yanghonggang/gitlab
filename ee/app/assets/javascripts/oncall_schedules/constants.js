export const LENGTH_ENUM = {
  hours: 'hours',
  days: 'days',
  weeks: 'weeks',
};

export const CHEVRON_SKIPPING_SHADE_ENUM = ['500', '600', '700', '800', '900', '950'];

export const CHEVRON_SKIPPING_PALETTE_ENUM = ['blue', 'orange', 'aqua', 'green', 'magenta'];
