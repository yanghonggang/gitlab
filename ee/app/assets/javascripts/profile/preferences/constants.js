import { s__ } from '~/locale';

export const i18n = {
  groupOverview: s__('ProfilePreferences|Group overview content'),
  groupOverviewDescription: s__(
    'ProfilePreferences|Choose what content you want to see on a groupâ€™s overview page.',
  ),
};
