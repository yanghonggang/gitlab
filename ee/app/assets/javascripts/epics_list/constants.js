import { __ } from '~/locale';
import { AvailableSortOptions } from '~/issuable_list/constants';

export const EpicsSortOptions = [
  {
    id: AvailableSortOptions.length + 1,
    title: __('Start date'),
    sortDirection: {
      descending: 'start_date_desc',
      ascending: 'start_date_asc',
    },
  },
  {
    id: AvailableSortOptions.length + 2,
    title: __('Due date'),
    sortDirection: {
      descending: 'end_date_desc',
      ascending: 'end_date_asc',
    },
  },
];

export const FilterTokenOperators = [
  { value: '=', description: __('is'), default: 'true' },
  // { value: '!=', description: __('is not') },
];
