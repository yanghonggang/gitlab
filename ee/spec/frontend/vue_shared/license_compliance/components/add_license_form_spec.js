import { GlFormRadio, GlFormRadioGroup } from '@gitlab/ui';
import { shallowMount, mount } from '@vue/test-utils';
import Vue from 'vue';
import LicenseIssueBody from 'ee/vue_shared/license_compliance/components/add_license_form.vue';
import { LICENSE_APPROVAL_STATUS } from 'ee/vue_shared/license_compliance/constants';

const KNOWN_LICENSES = [{ name: 'BSD' }, { name: 'Apache' }];

let wrapper;
let vm;

const createComponent = (props = {}, mountFn = shallowMount) => {
  wrapper = mountFn(LicenseIssueBody, { propsData: { knownLicenses: KNOWN_LICENSES, ...props } });
  vm = wrapper.vm;
};

describe('AddLicenseForm', () => {
  const findSubmitButton = () => wrapper.find('.js-submit');
  const findCancelButton = () => wrapper.find('.js-cancel');
  const findRadioInputs = () => wrapper.find(GlFormRadioGroup).findAll(GlFormRadio);

  beforeEach(() => {
    createComponent();
  });

  afterEach(() => {
    vm = undefined;
    wrapper.destroy();
  });

  describe('interaction', () => {
    it('clicking the Submit button submits the data and closes the form', async () => {
      const name = 'LICENSE_TEST';

      createComponent({}, mount);
      jest.spyOn(vm, '$emit').mockImplementation(() => {});
      wrapper.setData({ approvalStatus: LICENSE_APPROVAL_STATUS.ALLOWED, licenseName: name });

      await Vue.nextTick();

      const linkEl = findSubmitButton();
      linkEl.trigger('click');

      expect(vm.$emit).toHaveBeenCalledWith('addLicense', {
        newStatus: LICENSE_APPROVAL_STATUS.ALLOWED,
        license: { name },
      });
    });

    it('clicking the Cancel button closes the form', () => {
      createComponent({}, mount);
      const linkEl = findCancelButton();
      jest.spyOn(vm, '$emit').mockImplementation(() => {});
      linkEl.trigger('click');

      expect(vm.$emit).toHaveBeenCalledWith('closeForm');
    });
  });

  describe('computed', () => {
    describe('submitDisabled', () => {
      it('is true if the approvalStatus is empty', () => {
        wrapper.setData({ licenseName: 'FOO', approvalStatus: '' });

        expect(vm.submitDisabled).toBe(true);
      });

      it('is true if the licenseName is empty', () => {
        wrapper.setData({ licenseName: '', approvalStatus: LICENSE_APPROVAL_STATUS.ALLOWED });

        expect(vm.submitDisabled).toBe(true);
      });

      it('is true if the entered license is duplicated', () => {
        createComponent({ managedLicenses: [{ name: 'FOO' }] });
        wrapper.setData({ licenseName: 'FOO', approvalStatus: LICENSE_APPROVAL_STATUS.ALLOWED });

        expect(vm.submitDisabled).toBe(true);
      });
    });

    describe('isInvalidLicense', () => {
      it('is true if the entered license is duplicated', () => {
        createComponent({ managedLicenses: [{ name: 'FOO' }] });
        wrapper.setData({ licenseName: 'FOO' });

        expect(vm.isInvalidLicense).toBe(true);
      });

      it('is false if the entered license is unique', () => {
        createComponent({ managedLicenses: [{ name: 'FOO' }] });
        wrapper.setData({ licenseName: 'FOO2' });

        expect(vm.isInvalidLicense).toBe(false);
      });
    });
  });

  describe('template', () => {
    it('renders the license select dropdown', () => {
      const dropdownElement = wrapper.find('#js-license-dropdown');

      expect(dropdownElement.exists()).toBe(true);
    });

    it('renders the license approval radio buttons dropdown', () => {
      const approvalOptions = findRadioInputs();

      expect(approvalOptions).toHaveLength(2);
      expect(approvalOptions.at(0).text()).toBe('Allow');
      expect(approvalOptions.at(1).text()).toBe('Deny');
    });

    it('renders error text, if there is a duplicate license', async () => {
      createComponent({ managedLicenses: [{ name: 'FOO' }] });
      wrapper.setData({ licenseName: 'FOO' });
      await Vue.nextTick();

      const feedbackElement = wrapper.find('.invalid-feedback');

      expect(feedbackElement.exists()).toBe(true);
      expect(feedbackElement.classes()).toContain('d-block');
      expect(feedbackElement.text()).toBe('This license already exists in this project.');
    });

    it('shows radio button descriptions, if licenseComplianceDeniesMr feature flag is enabled', async () => {
      wrapper = shallowMount(LicenseIssueBody, {
        propsData: {
          managedLicenses: [{ name: 'FOO' }],
          knownLicenses: KNOWN_LICENSES,
        },
        provide: {
          glFeatures: { licenseComplianceDeniesMr: true },
        },
      });

      await Vue.nextTick();

      const descriptionElement = wrapper.findAll('.text-secondary');

      expect(descriptionElement.at(0).text()).toBe('Acceptable license to be used in the project');

      expect(descriptionElement.at(1).text()).toBe(
        'Disallow merge request if detected and will instruct developer to remove',
      );
    });

    it('does not show radio button descriptions, if licenseComplianceDeniesMr feature flag is disabled', () => {
      createComponent({ managedLicenses: [{ name: 'FOO' }] });
      wrapper.setData({ licenseName: 'FOO' });
      return Vue.nextTick().then(() => {
        expect(findRadioInputs().at(0).element).toMatchSnapshot();
        expect(findRadioInputs().at(1).element).toMatchSnapshot();
      });
    });

    it('disables submit, if the form is invalid', async () => {
      wrapper.setData({ licenseName: '' });
      await Vue.nextTick();

      expect(vm.submitDisabled).toBe(true);

      const submitButton = findSubmitButton();

      expect(submitButton.exists()).toBe(true);
      expect(submitButton.props().disabled).toBe(true);
    });

    it('disables submit and cancel while a new license is being added', async () => {
      wrapper.setProps({ loading: true });
      await Vue.nextTick();

      const submitButton = findSubmitButton();
      const cancelButton = findCancelButton();

      expect(submitButton.exists()).toBe(true);
      expect(submitButton.props().disabled).toBe(true);
      expect(cancelButton.exists()).toBe(true);
      expect(cancelButton.props().disabled).toBe(true);
    });
  });
});
