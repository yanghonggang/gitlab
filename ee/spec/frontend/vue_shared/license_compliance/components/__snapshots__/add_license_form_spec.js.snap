// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddLicenseForm template does not show radio button descriptions, if licenseComplianceDeniesMr feature flag is disabled 1`] = `
<gl-form-radio-stub
  aria-describedby="js-approved-license-radio"
  class=""
  data-qa-selector="approved_license_radio"
  value="approved"
>
  Allow 
  <!---->
</gl-form-radio-stub>
`;

exports[`AddLicenseForm template does not show radio button descriptions, if licenseComplianceDeniesMr feature flag is disabled 2`] = `
<gl-form-radio-stub
  aria-describedby="js-blacklisted-license-radio"
  class=""
  data-qa-selector="blacklisted_license_radio"
  value="blacklisted"
>
  Deny 
  <!---->
</gl-form-radio-stub>
`;
